src/main/java/com/cleanroommc/groovyscript/compat/mods/advancedmortars/Mortar.java

```diff
@@ -5,54 +5,60 @@
 import com.cleanroommc.groovyscript.helper.ingredient.IngredientHelper;
+import com.cleanroommc.groovyscript.helper.recipe.AbstractRecipeBuilder;
 import com.cleanroommc.groovyscript.registry.VirtualizedRegistry;
 import com.codetaylor.mc.advancedmortars.modules.mortar.api.MortarAPI;
 import com.codetaylor.mc.advancedmortars.modules.mortar.recipe.RecipeMortar;
 import com.codetaylor.mc.advancedmortars.modules.mortar.reference.EnumMortarType;
 import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.Nullable;
 
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
 
 public class Mortar extends VirtualizedRegistry<RecipeMortar> {
 
     public Mortar() {
         super();
     }
 
+    public RecipeBuilder recipeBuilder() {
+        return new RecipeBuilder();
+    }
+
     @Override
     @GroovyBlacklist
     public void onReload() {
         removeScripted().forEach(this::remove);
         restoreFromBackup().forEach(recipe -> getTypes(recipe).forEach(type -> add(type, recipe)));
     }
 
-    public void add(String[] types, ItemStack output, int duration, List<IIngredient> inputs) {
+    public void add(List<String> types, ItemStack output, int duration, List<IIngredient> inputs) {
         add(types, output, duration, ItemStack.EMPTY, 0.0f, inputs);
     }
 
-    public void add(String[] types, ItemStack output, int duration, ItemStack secondaryOutput, float secondaryOutputChance, List<IIngredient> inputs) {
-        if (inputs != null && inputs.size() != 0) {
-            if (inputs.size() > 8) {
-                GroovyLog.msg("Error adding Advanced Mortars recipe")
-                        .add("maximum number of 8 input ingredients exceeded: " + inputs.size())
-                        .error()
-                        .post();
-            } else {
-                for (String type : types) {
-                    EnumMortarType enumMortarType = EnumMortarType.fromName(type);
-                    if (enumMortarType == null) {
-                        GroovyLog.msg("Error adding Advanced Mortars recipe")
-                                .add("invalid mortar type: " + type)
-                                .add("valid types are: " + Arrays.toString(EnumMortarType.NAMES))
-                                .error()
-                                .post();
-                    } else {
-                        add(enumMortarType, new RecipeMortar(
-                                output, duration, secondaryOutput, secondaryOutputChance, IngredientHelper.toIngredientNonNullList(inputs)
-                        ));
-                    }
+    public void add(List<String> types, ItemStack output, int duration, ItemStack secondaryOutput, float secondaryOutputChance, List<IIngredient> inputs) {
+        if (inputs == null || inputs.size() == 0) return;
+        if (inputs.size() > 8) {
+            GroovyLog.msg("Error adding Advanced Mortars recipe")
```

Comment:
since this will generate an error and not add a recipe, you should return early here too.